language-plutus-core/src/Language/PlutusCore/Renamer.hs

```diff
@@ -111,104 +112,79 @@ annotateTy (TyApp x ty ty') =
-renameTerm st (Apply x t t') = Apply x <$> renameTerm st t <*> renameTerm st t'
-renameTerm st (Unwrap x t) = Unwrap x <$> renameTerm st t
-renameTerm _ x@Constant{} = pure x
-renameTerm st (Error x ty) = Error x <$> renameType st ty
-renameTerm st (TyInst x t tys) = TyInst x <$> renameTerm st t <*> renameType st tys
-
-renameType :: Identifiers -> Type TyName a -> MaxM (Type TyName a)
-renameType _ ty@TyInt{} = pure ty
-renameType st ty@(TyLam x (TyName (Name x' s (Unique u))) k ty') = do
-    m <- get
-    let st' = modifyIdentifiers u m st
-        pastDef = lookupId u st
-    case pastDef of
-        Just _ ->
-            modify (+1) >>
-            TyLam x (TyName (Name x' s (Unique (m+1)))) k <$> renameType st' ty'
-        _ -> renameType st' ty
-renameType st ty@(TyForall x (TyName (Name x' s (Unique u))) k ty') = do
-    m <- get
-    let st' = modifyIdentifiers u m st
-        pastDef = lookupId u st
-    case pastDef of
-        Just _ ->
-            modify (+1) >>
-            TyForall x (TyName (Name x' s (Unique (m+1)))) k <$> renameType st' ty'
-        _ -> renameType st' ty
-renameType st ty@(TyFix x (TyName (Name x' s (Unique u))) ty') = do
-    m <- get
-    let st' = modifyIdentifiers u m st
-        pastDef = lookupId u st
-    case pastDef of
-        Just _ ->
-            modify (+1) >>
-            TyFix x (TyName (Name x' s (Unique (m+1)))) <$> renameType st' ty'
-        _ -> renameType st' ty
-renameType st ty@(TyVar x (TyName (Name x' s (Unique u)))) =
-    case pastDef of
-        Just j -> pure $ TyVar x (TyName (Name x' s (Unique j)))
-        _      -> pure ty
-    where pastDef = lookupId u st
-renameType st (TyApp x ty ty') = TyApp x <$> renameType st ty <*> renameType st ty'
-renameType st (TyFun x ty ty') = TyFun x <$> renameType st ty <*> renameType st ty'
-renameType _ ty@TyBuiltin{} = pure ty
+-- | Mapping from locally unique indices to globally unique 'Unique's.
+newtype UniquesRenaming = UniquesRenaming
+    { unUniquesRenaming :: IM.IntMap Unique
+    }
+
+-- | The monad the renamer runs in.
+type RenameM = ReaderT UniquesRenaming Quote
+
+-- | Rename 'Unique's so that they're globally unique.
+rename :: (Rename a, MonadQuote m) => a -> m a
+rename x = liftQuote $ runReaderT (renameM x) (UniquesRenaming mempty)
+
+-- | Save the mapping from an old 'Unique' to a new one.
+updateUniquesRenaming :: Unique -> Unique -> UniquesRenaming -> UniquesRenaming
+updateUniquesRenaming uniqOld uniqNew (UniquesRenaming uniques) =
+    UniquesRenaming $ IM.insert (unUnique uniqOld) uniqNew uniques
+
+-- | Look up a new 'Unique' an old 'Unique' got mapped to.
+lookupUnique :: Unique -> RenameM (Maybe Unique)
+lookupUnique uniq = asks $ IM.lookup (unUnique uniq) . unUniquesRenaming
+
+-- | Replace the 'Unique' in a value by a new 'Unique', save the mapping
+-- from an old 'Unique' to the new one and supply the updated value to a continuation.
+withRefreshed :: HasUnique a => a -> (a -> RenameM c) -> RenameM c
```

Comment:
Why is this generic using `HasUnique`? We don't ever want to rename a `Type TyNameWithKind a` do we?

Reply:
Just so that it can be applied to both `TyName` and `Name`.